rmw_fastrtps_shared_cpp/src/rmw_take.cpp

```diff
@@ -353,4 +360,133 @@ __rmw_take_serialized_message_with_info(
   return _take_serialized_message(
     identifier, subscription, serialized_message, taken, message_info, allocation);
 }
+
+// ----------------- Loans related code ------------------------- //
+
+struct GenericSequence : public eprosima::fastdds::dds::LoanableCollection
+{
+  GenericSequence() = default;
+
+  void resize(size_type /*new_length*/) override
+  {
+    // This kind of collection should only be used with loans
+    throw std::bad_alloc();
+  }
+};
+
+struct LoanManager
+{
+  struct Item
+  {
+    GenericSequence data_seq{};
+    eprosima::fastdds::dds::SampleInfoSeq info_seq{};
+  };
+
+  explicit LoanManager(const eprosima::fastrtps::ResourceLimitedContainerConfig & items_cfg)
+  : items(items_cfg)
+  {
+  }
+
+  std::mutex mtx;
+  eprosima::fastrtps::ResourceLimitedVector<Item> items RCPPUTILS_TSA_GUARDED_BY(mtx);
+};
+
+void
+__init_subscription_for_loans(
+  rmw_subscription_t * subscription)
+{
+  auto info = static_cast<CustomSubscriberInfo *>(subscription->data);
+  const auto & qos = info->data_reader_->get_qos();
+  bool has_data_sharing = DataSharingKind::OFF != qos.data_sharing().kind();
+  subscription->can_loan_messages = has_data_sharing && info->type_support_->is_plain();
+  if (subscription->can_loan_messages) {
+    const auto & allocation_qos = qos.reader_resource_limits().outstanding_reads_allocation;
+    info->loan_manager_ = std::make_shared<LoanManager>(allocation_qos);
+  }
+}
+
+rmw_ret_t
+__rmw_take_loaned_message_internal(
+  const char * identifier,
+  const rmw_subscription_t * subscription,
+  void ** loaned_message,
+  bool * taken,
+  rmw_message_info_t * message_info)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(subscription, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    subscription, subscription->implementation_identifier, identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  if (!subscription->can_loan_messages) {
+    RMW_SET_ERROR_MSG("Loaning is not supported");
+    return RMW_RET_UNSUPPORTED;
+  }
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(loaned_message, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(taken, RMW_RET_INVALID_ARGUMENT);
+
+  auto info = static_cast<CustomSubscriberInfo *>(subscription->data);
+  auto loan_mgr = info->loan_manager_;
+  std::unique_lock<std::mutex> guard(loan_mgr->mtx);
+  auto item = loan_mgr->items.emplace_back();
+  if (nullptr == item) {
+    RMW_SET_ERROR_MSG("Out of resources for loaned message info");
+    return RMW_RET_ERROR;
```

Comment:
@MiguelCompany nit: `RMW_RET_BAD_ALLOC`?

Reply:
Done in 1332adb6d2704e56e476d2182a5c458599785911